File: service/src/main/java/net/huray/backend/minuting/error/GlobalExceptionHandler.java
package net.huray.backend.minuting.error;

import feign.RetryableException;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

import static net.huray.backend.minuting.error.ErrorCode.External_Server_Error;
import static net.huray.backend.minuting.error.ErrorCode.Internal_Server_Error;

@ControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleException(final Exception e) {
        e.printStackTrace();
        return getResponse(Internal_Server_Error);
    }

    @ExceptionHandler(RetryableException.class)
    public ResponseEntity<ErrorResponse> handleRetryableException(final RetryableException e) {
        return getResponse(External_Server_Error);
    }

    @ExceptionHandler(GlobalException.class)
    public ResponseEntity<ErrorResponse> handleGlobalException(final GlobalException e) {
        return getResponse(e.getStatus(), e.getMessage());
    }

    private ResponseEntity<ErrorResponse> getResponse(final ErrorCode code) {
        final ErrorResponse response = new ErrorResponse(code.getMessage());
        return new ResponseEntity<>(response, code.getStatus());
    }

    private ResponseEntity<ErrorResponse> getResponse(final HttpStatus code, final String message) {
        final ErrorResponse response = new ErrorResponse(message);
        return new ResponseEntity<>(response, code);
    }

}

File: service/src/main/kotlin/net/huray/backend/minuting/support/GlobalExceptionHandler.kt
package net.huray.backend.minuting.support

import feign.RetryableException
import net.huray.backend.http.exception.BaseException
import net.huray.backend.http.res.BaseError
import net.huray.backend.http.res.BaseResult
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.INTERNAL_SERVER_ERROR
import org.springframework.http.HttpStatus.SERVICE_UNAVAILABLE
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<BaseResult> {
        e.printStackTrace()
        return ResponseEntity(BaseResult(BaseError(500, "Internal Server Error")), INTERNAL_SERVER_ERROR)
    }

    @ExceptionHandler(RetryableException::class)
    fun handleRetryableException(e: RetryableException): ResponseEntity<BaseResult> =
        ResponseEntity(BaseResult(BaseError(e.status(), "External Server Error")), SERVICE_UNAVAILABLE)

    @ExceptionHandler(BaseException::class)
    fun handleBaseException(e: BaseException): ResponseEntity<BaseResult> =
        ResponseEntity(BaseResult(BaseError(e.code, e.reason)), HttpStatus.valueOf(e.code))

}